### DISCUSSION TITLE
Unable to customize Decimal serialization (Annotated fields can't work)

### DISCUSSION THREAD

**Discussion by MrNaif2018** [fastapi/fastapi#9983] [Questions] [answered]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
from decimal import Decimal
from typing import Union

from fastapi import FastAPI
from pydantic import BaseModel, SerializerFunctionWrapHandler, WrapSerializer
from typing_extensions import Annotated


def decimal_encoder(dec_value: Decimal, next: SerializerFunctionWrapHandler) -> Union[int, float]:
    if dec_value.as_tuple().exponent >= 0:
        return int(dec_value)
    else:
        return float(dec_value)


Decimal = Annotated[Decimal, WrapSerializer(decimal_encoder)]


class Model(BaseModel):
    field1: Decimal
    field2: dict


app = FastAPI()


@app.get("/", response_model=Model)
def root():
    return {"field1": Decimal("1.4600000000"), "field2": {"a": 1, "b": 2, "c": {"val": Decimal("1.4600000000")}}}
```

### Description

I am able to customize Decimal serialization to work like in fastapi used with pydantic v1 when the field is a direct field on a model
But in case I have some custom possibly untyped fields, it defaults to built-in serialization and there seems to be no way to override it. json_encoders on BaseModel also don't work for some reason
field1 gets serialized properly, but field2 is serialized as a full string
I can't define exact models because I have a plugin system where users may add custom metadata

### Operating System

Linux Mint

### FastAPI Version

0.100.1

### Python Version

3.8

**Comment by MrNaif2018** [comment]
@tiangolo @Kludex @dmontagu Sorry for the ping, just that found out this issue also impacts other places like datetime formatting
In pydantic v2 it now uses Zulu format which isn't parseable by datetime.fromisoformat, and the solution provided here https://github.com/pydantic/pydantic/issues/6028 does not work either because a datetime which is in i.e. untyped dict can't be annotated and gets converted to incompatible format

**Reply by botsman** [reply]
I have the same issue: untyped response containing Decimal type is converted to float.
I would like to customise that behaviour and return it as string

**Reply by romabozhanovgithub** [reply]
Also tried it with `return_type`, didn't work
```python
from decimal import Decimal
from typing import Union

from fastapi import FastAPI
from pydantic import BaseModel, SerializerFunctionWrapHandler, WrapSerializer
from typing_extensions import Annotated


def decimal_encoder(dec_value: Decimal, next: SerializerFunctionWrapHandler) -> Union[int, float]:
    print("decimal_encoder", dec_value)
    if dec_value.as_tuple().exponent >= 0:
        return int(dec_value)
    else:
        return float(dec_value)


Decimal = Annotated[Decimal, WrapSerializer(decimal_encoder, return_type=Union[int, float])]


class Model(BaseModel):
    field1: Decimal
    field2: dict


app = FastAPI()


@app.get("/", response_model=Model)
def root():
    return {
        "field1": Decimal("1.4600000000"),
        "field2": {
            "a": 1,
            "b": 2,
            "c": {
                "val": Decimal("1.46000000000")
            }
        }
    }

```

**Reply by MrNaif2018** [reply]
Exactly. It doesn't work because no matter how we annotate Decimal, we can't modify internal serialization outside of pydantic models

**Reply by botsman** [reply]
I've also tried doing the following:
```python
from typing import Any
from decimal import Decimal

from fastapi.responses import JSONResponse
from fastapi.encoders import jsonable_encoder

class CustomJSONResponse(JSONResponse):
    def render(self, content: Any) -> bytes:
        return super().render(jsonable_encoder(content, custom_encoder={Decimal: str}))
        
        
@app.get("/some", response_class=CustomJSONResponse)
async def some():
    return {"d": Decimal("123.456")}
```

But it appears that the response which is passed to the jsonable_encoder already has Decimal converted to float.

**Comment by iloveitaly** [comment]
@MrNaif2018 did you ever figure out a solution here?

**Reply by MrNaif2018** [reply]
I use those 2 types, don't remember if it changed API behaviour or not:
```
NonZuluDatetime = Annotated[datetime, PlainSerializer(lambda v: v.isoformat(), return_type=str, when_used="json")]
DecimalAsFloat = Annotated[Decimal, PlainSerializer(lambda v: float(v), return_type=float, when_used="json")]
```

**Comment by YuriiMotov** [comment, marked as answer]
This works:

```python
from decimal import Decimal
from typing import Union

from fastapi import FastAPI
from fastapi.encoders import jsonable_encoder
from pydantic import BaseModel, PlainSerializer
from typing_extensions import Annotated


def decimal_encoder(dec_value: Decimal) -> Union[int, float]:
    if dec_value.as_tuple().exponent >= 0:
        return int(dec_value)
    else:
        return float(dec_value)


MyDecimal = Annotated[Decimal, PlainSerializer(decimal_encoder)]


class Model(BaseModel):
    field1: MyDecimal
    field2: dict


app = FastAPI()


@app.get("/", response_model=Model)
def root():
    res = {
        "field1": Decimal("1.4600000000"),
        "field2": {"a": 1, "b": 2, "c": {"val": Decimal("1.4600000000")}},
    }
    return jsonable_encoder(res, custom_encoder={Decimal: decimal_encoder})

```

Note, that when using Pydantic V2, you need to specify encoder both in `jsonable_encoder`'s parameter and in model.
`jsonable_encoder` doesn't affect the serialization of Pydantic V2 models.

Response:
```json
{
  "field1": 1.46,
  "field2": {
    "a": 1,
    "b": 2,
    "c": {
      "val": 1.46
    }
  }
}
```
